File: Huffman Encoding/huffman.cpp
#include<bits/stdc++.h>
using namespace std;

class Node {
public:
    int val;
    string symbol;
    string huffman;
    Node *left;
    Node *right;

    Node(int n, string s) {
        this -> val = n;
        this -> symbol = s;
        this -> huffman = "";
        this -> left = nullptr;
        this -> right = nullptr;
    }
};


struct Compare {
    bool operator()(const Node *a, const Node *b) {
        return a -> val > b -> val;
    }
};


class HuffmanEncoding {
    Node *root;
    string inputText;
    unordered_map<string, int> freqs;
    unordered_map<string, string> codes;
    priority_queue<Node *, vector<Node *>, Compare> pq;

    void encode(Node *node, const string &val = "") {
        string newVal = val + node -> huffman;

        if(node -> left != nullptr) {
            encode(node -> left, newVal);
        }
        if(node -> right != nullptr) {
            encode(node -> right, newVal);
        }

        if(node -> left == nullptr && node -> right == nullptr) {
            codes[node -> symbol] = newVal;
        }
    }

public:
    HuffmanEncoding(string text) {
        this -> inputText = text;
        this -> root = nullptr;

        for(char c : inputText) {
            string s(1, c);
            freqs[s]++;
        }

        for(auto it : freqs) {
            Node *node = new Node(it.second, it.first);
            pq.push(node);
        }
    }

    void buildTree() {
        while(pq.size() > 1) {
            Node *l = pq.top();
            pq.pop();
            Node *r = pq.top();
            pq.pop();

            l -> huffman = "0";
            r -> huffman = "1";

            Node *node = new Node(l -> val + r -> val, l -> symbol + r -> symbol);
            node -> left = l;
            node -> right = r;
            pq.push(node);
        }

        this -> root = pq.top();
        pq.pop();
    }

    void displayResult() {
        cout<<"The character coding is as follows: "<<endl;
        encode(root);
        for(auto it : codes) {
            cout<<it.first<<" --> "<<it.second<<endl;
        }
        cout<<endl;

        string encodedText = "";
        for(char c : inputText) {
            string s(1, c);
            encodedText += codes[s];
        }

        cout<<"The encoded text is.."<<endl;
        cout<<encodedText<<endl;
    }
};


int main() {
    string text;
    cout<<"Enter text to encode"<<endl;
    cin>>text;

    HuffmanEncoding he(text);
    he.buildTree();
    he.displayResult();

    return 0;
}

File: Huffman Encoding/huffman.py
import heapq

class Node:
    def __init__(self, n, s) -> None:
        self.val = n
        self.symbol = s
        self.huffman = ''
        self.left = None
        self.right = None

    def __lt__(self, other):
        return self.val < other.val
    

class HuffmanEncoding:
    def __init__(self, text) -> None:
        self.root = None
        self.input_text = text
        self.codes = {}
        self.pq = []

        self.freqs = {}
        for char in self.input_text:
            if char in self.freqs:
                self.freqs[char] += 1
            else:
                self.freqs[char] = 1

        for key, value in self.freqs.items():
            node = Node(value, key)
            heapq.heappush(self.pq, node)

    def encode(self, node: Node, val=""):
        new_val = val + node.huffman

        if node.left:
            self.encode(node.left, new_val)
        if node.right:
            self.encode(node.right, new_val)

        if not node.left and not node.right:
            self.codes[node.symbol] = new_val

    def build_tree(self):
        while len(self.pq) > 1:
            l = heapq.heappop(self.pq)
            r = heapq.heappop(self.pq)

            l.huffman = '0'
            r.huffman = '1'

            node = Node(l.val + r.val, l.symbol + r.symbol)
            node.left = l
            node.right = r
            heapq.heappush(self.pq, node)

        self.root = heapq.heappop(self.pq)

    def display_result(self):
        print("The character coding is as follows:")
        self.encode(self.root)
        for key, value in self.codes.items():
            print(key + " --> " + value)        

        encoded_text = ''
        for char in self.input_text:
            encoded_text += self.codes.get(char, 0)
        
        print("The encoded text is...")
        print(encoded_text)


text = input("Enter text to encode")
he = HuffmanEncoding(text)
he.build_tree()
he.display_result()